File: algorithm/recursion/Code10-14.cpp
#include <iostream>
#include <random>

// 색상 값을 함께 출력해 전체 원 프랙탈의 호출 구조를 확인합니다.
void drawColorCircle(int x, int y, int r, std::mt19937& gen) {
    std::uniform_int_distribution<int> color(0, 255);
    std::cout << "circle: center=(" << x << ", " << y << "), radius=" << r
              << ", RGB=(" << color(gen) << ", " << color(gen) << ", " << color(gen) << ")\n";
    if (r >= 5) {
        drawColorCircle(x - r / 2, y, r / 2, gen);
        drawColorCircle(x + r / 2, y, r / 2, gen);
    }
}

int main() {
    std::mt19937 gen(std::random_device{}());
    drawColorCircle(64, 64, 64, gen);
}

File: algorithm/recursion/Code10-14.py
import matplotlib.pyplot as plt
from matplotlib.patches import Circle
import random

def draw_color_circle(ax, x, y, r):
    # 무작위 RGB 색상 만들기
    color = (
        random.randint(0, 255) / 255,
        random.randint(0, 255) / 255,
        random.randint(0, 255) / 255
    )

    # 실제 원 그리기
    circle = Circle(
        (x, y),
        r,
        fill=False,
        edgecolor=color,
        linewidth=2
    )
    ax.add_patch(circle)

    # 반지름이 5 이상이면 좌우에 작은 원을 재귀로 그림
    if r >= 5:
        draw_color_circle(ax, x - r / 2, y, r / 2)
        draw_color_circle(ax, x + r / 2, y, r / 2)


fig, ax = plt.subplots(figsize=(12, 4))

draw_color_circle(ax, 64, 64, 64)

ax.set_xlim(0, 128)
ax.set_ylim(0, 128)
ax.set_aspect("equal")
ax.set_facecolor("black")
ax.axis("off")

plt.show()